crates/codec/src/api/wasm/buffer.rs

```diff
@@ -0,0 +1,258 @@
+#[derive(Debug, PartialEq)]
+pub struct Buffer(Option<Vec<u8>>, bool);
+
+impl Buffer {
+    const LEN: Range<usize> = 0..4;
+    const CAPACITY: Range<usize> = 4..8;
+    const DATA: RangeFrom<usize> = 8..;
+
+    const HEADER_SIZE: usize = 8;
+
+    const OK_MARKER: u8 = 1;
+    const ERR_MARKER: u8 = 0;
+
+    pub fn alloc(len: u32) -> Self {
+        let mut vec = vec![0; len as usize + Self::HEADER_SIZE];
+        vec.shrink_to_fit();
+
+        assert_eq!(vec.capacity(), len as usize + Self::HEADER_SIZE);
+        assert_eq!(vec.len(), len as usize + Self::HEADER_SIZE);
+
+        let mut buf = Self(Some(vec), true);
+
+        buf.set_capacity(len);
+        buf.set_len(len);
+
+        buf
+    }
+
+    #[cfg(test)]
+    pub fn then_free(mut self) -> Self {
+        self.1 = false;
+        self
+    }
+
+    pub fn alloc_ok(data: &[u8]) -> Self {
+        let len = u32::try_from(data.len()).unwrap() + 1u32;
+        let mut buf = Self::alloc(len);
+
+        buf.as_mut()[0] = Self::OK_MARKER;
+        buf.as_mut()[1..].clone_from_slice(data);
+
+        buf
+    }
+
+    #[allow(unused)]
+    pub fn as_result(&self) -> Option<Result<&[u8], &str>> {
+        let first_byte = self.as_ref().get(0)?;
+        match *first_byte {
+            Self::OK_MARKER => Some(Ok(&self.as_ref()[1..])),
+            Self::ERR_MARKER => {
+                let s = std::str::from_utf8(&self.as_ref()[1..]).ok()?;
+                Some(Err(s))
+            }
+            _ => None,
+        }
+    }
+
+    pub fn alloc_err(string: impl ToString) -> Self {
+        let string = string.to_string();
+        let len = u32::try_from(string.as_bytes().len()).unwrap() + 1u32;
+        let mut buf = Self::alloc(len);
+
+        buf.as_mut()[0] = Self::ERR_MARKER;
+        buf.as_mut()[1..].clone_from_slice(string.as_bytes());
+
+        buf
+    }
+
+    pub fn free(mut self) {
+        self.0.take();
+    }
+
+    pub fn ptr(&mut self) -> *mut u8 {
+        self.vec_mut().as_mut_ptr()
+    }
+
+    pub unsafe fn from_ptr(ptr: *mut u8) -> Self {
+        let header = Self(
+            Some(Vec::from_raw_parts(
+                ptr,
+                Self::HEADER_SIZE,
+                Self::HEADER_SIZE,
+            )),
+            true,
+        );
+
+        Self(
+            Some(Vec::from_raw_parts(
+                ptr,
+                header.len() as usize + Self::HEADER_SIZE,
+                header.capacity() as usize + Self::HEADER_SIZE,
+            )),
+            true,
+        )
+    }
+
+    fn vec(&self) -> &Vec<u8> {
+        self.0.as_ref().unwrap()
+    }
+
+    fn vec_mut(&mut self) -> &mut Vec<u8> {
+        self.0.as_mut().unwrap()
+    }
+
+    pub fn len(&self) -> u32 {
+        u32::from_be_bytes((&self.vec()[Self::LEN]).try_into().unwrap())
+    }
+
+    pub fn capacity(&self) -> u32 {
+        u32::from_be_bytes((&self.vec()[Self::CAPACITY]).try_into().unwrap())
+    }
+
+    pub fn set_len(&mut self, len: u32) {
```

Comment:
I don't think we need `set_len` to be public. The current implementation just allocates a buffer once.
Why can't we just have the `alloc_ok / alloc_err` above?

Reply:
Yup, I agree!